File: apps/apple-native/Sources/ACoolProfile/ACoolAudioCompanion.swift
import Foundation

public enum ACoolAudioSourceMode: String, Codable, Sendable {
    case acoolOwned = "acool_owned"
    case userOwned = "user_owned"
    case licensedProvider = "licensed_provider"
    case externalDeepLink = "external_deep_link"
    case none
}

public struct ACoolAudioPolicy: Codable, Equatable, Sendable {
    public var sourceMode: ACoolAudioSourceMode
    public var musicEnabled: Bool
    public var duckForNavigation: Bool
    public var pauseForSafetyAlerts: Bool
    public var rightsVerified: Bool
    public var providerApproved: Bool

    public init(
        sourceMode: ACoolAudioSourceMode = .none,
        musicEnabled: Bool = false,
        duckForNavigation: Bool = true,
        pauseForSafetyAlerts: Bool = true,
        rightsVerified: Bool = false,
        providerApproved: Bool = false
    ) {
        self.sourceMode = sourceMode
        self.musicEnabled = musicEnabled
        self.duckForNavigation = duckForNavigation
        self.pauseForSafetyAlerts = pauseForSafetyAlerts
        self.rightsVerified = rightsVerified
        self.providerApproved = providerApproved
    }

    public var canStartInAppPlayback: Bool {
        guard musicEnabled else { return false }
        switch sourceMode {
        case .acoolOwned, .userOwned:
            return rightsVerified
        case .licensedProvider:
            return rightsVerified && providerApproved
        case .externalDeepLink, .none:
            return false
        }
    }

    public func shouldDuckForGuidance(isNavigationActive: Bool) -> Bool {
        isNavigationActive && duckForNavigation
    }

    public func shouldPauseForAlert(isSafetyAlert: Bool) -> Bool {
        isSafetyAlert && pauseForSafetyAlerts
    }
}

File: apps/android-native/src/main/kotlin/com/acoolcollector/nativeapp/media/ACoolAudioCompanion.kt
package com.acoolcollector.nativeapp.media

enum class ACoolAudioSourceMode {
    ACOOL_OWNED,
    USER_OWNED,
    LICENSED_PROVIDER,
    EXTERNAL_DEEP_LINK,
    NONE,
}

data class ACoolAudioPolicy(
    val sourceMode: ACoolAudioSourceMode = ACoolAudioSourceMode.NONE,
    val musicEnabled: Boolean = false,
    val duckForNavigation: Boolean = true,
    val pauseForSafetyAlerts: Boolean = true,
    val rightsVerified: Boolean = false,
    val providerApproved: Boolean = false,
) {
    fun canStartInAppPlayback(): Boolean {
        if (!musicEnabled) return false
        return when (sourceMode) {
            ACoolAudioSourceMode.ACOOL_OWNED,
            ACoolAudioSourceMode.USER_OWNED -> rightsVerified
            ACoolAudioSourceMode.LICENSED_PROVIDER -> rightsVerified && providerApproved
            ACoolAudioSourceMode.EXTERNAL_DEEP_LINK,
            ACoolAudioSourceMode.NONE -> false
        }
    }

    fun shouldDuckForGuidance(isNavigationActive: Boolean): Boolean =
        isNavigationActive && duckForNavigation

    fun shouldPauseForAlert(isSafetyAlert: Boolean): Boolean =
        isSafetyAlert && pauseForSafetyAlerts
}
